File: tasks/seeds/py-minesweeper/files/test_minesweeper.py
"""Acceptance tests for the minesweeper engine + CLI. Run: python3 test_minesweeper.py"""
import subprocess
import sys

LAYOUT1 = "layout1.txt"   # 6x6, mines at (1,1) (2,2) (3,4) (5,1)
LAYOUT2 = "layout2.txt"   # 4x4, mines at (0,0) (2,2)


def run_cli(layout, script):
    p = subprocess.run([sys.executable, "minesweeper.py", layout],
                       input=script, capture_output=True, text=True, timeout=60)
    assert p.returncode == 0, (p.returncode, p.stderr)
    return p.stdout


def expect_raises(fn, *args):
    try:
        fn(*args)
    except ValueError:
        return
    raise AssertionError(f"{fn.__name__}{args} should raise ValueError")


def main():
    from minesweeper import Game, load

    # -- layout validation --
    expect_raises(Game, [])
    expect_raises(Game, ["..", "..."])       # ragged
    expect_raises(Game, ["._"])              # unknown char
    expect_raises(Game, [""])                # zero-width

    # -- basic shape and counters --
    g = load(LAYOUT1)
    assert (g.rows, g.cols, g.mine_count) == (6, 6, 4)
    assert g.state == "playing"
    assert g.flags_left == 4

    # -- adjacency counts across the whole grid --
    expected = ["111000",
                "112100",
                "121211",
                "011201",
                "111111",
                "101000"]
    got = ["".join(str(g.adjacent(r, c)) for c in range(6)) for r in range(6)]
    assert got == expected, got

    # -- everything starts hidden --
    assert g.render() == "\n".join(["######"] * 6)

    # -- out-of-range coordinates are rejected --
    expect_raises(g.adjacent, -1, 0)
    expect_raises(g.reveal, 0, 6)
    expect_raises(g.flag, 6, 0)

    # -- revealing a numbered cell reveals only that cell --
    g = load(LAYOUT1)
    g.reveal(0, 0)
    assert g.render() == ("1#####\n######\n######\n######\n######\n######")
    assert g.state == "playing"

    # -- flood fill opens the connected zero region plus its numbered rim --
    g = load(LAYOUT1)
    g.reveal(0, 5)
    assert g.render() == ("##1...\n##21..\n###211\n######\n######\n######")
    assert len(g.revealed) == 11
    g2 = load(LAYOUT1)
    g2.reveal(5, 5)
    assert g2.render() == ("######\n######\n######\n######\n##1111\n##1...")
    assert len(g2.revealed) == 8

    # -- flags: toggle, block reveals, and are skipped by flood fill --
    g = load(LAYOUT1)
    g.flag(0, 4)
    assert g.flags_left == 3
    g.reveal(0, 4)                       # flagged: reveal is a no-op
    assert (0, 4) not in g.revealed
    g.reveal(0, 5)                       # flood must flow around the flag
    assert g.render() == ("##1.F.\n##21..\n###211\n######\n######\n######")
    assert len(g.revealed) == 10
    g.flag(0, 4)                         # unflag; still unrevealed
    assert g.flags_left == 4
    assert (0, 4) not in g.revealed
    g.reveal(0, 4)
    assert (0, 4) in g.revealed
    g.flag(0, 4)                         # flagging a revealed cell: no-op
    assert g.flags_left == 4
    g.flag(3, 0)
    g.flag(3, 1)
    g.flag(3, 2)
    g.flag(3, 3)
    g.flag(3, 5)
    assert g.flags_left == -1            # over-flagging just goes negative

    # -- stepping on a mine loses and unmasks every mine --
    g = load(LAYOUT1)
    g.reveal(0, 0)
    g.reveal(1, 1)
    assert g.state == "lost"
    assert g.render() == ("1#####\n#*####\n##*###\n####*#\n######\n#*####")
    before = set(g.revealed)
    g.reveal(4, 4)                       # game over: further moves are no-ops
    g.flag(4, 4)
    assert g.revealed == before and not g.flags

    # -- winning = every safe cell revealed; flags are irrelevant --
    g = load(LAYOUT2)
    for r, c in [(0, 3), (1, 0), (2, 0), (2, 3), (3, 3), (3, 2)]:
        assert g.state == "playing"
        g.reveal(r, c)
    assert g.state == "won"
    assert len(g.revealed) == 14

    # -- CLI: full winning transcript with a flag and a rejected command --
    assert run_cli(LAYOUT2, "r 0 3\nf 2 2\nr 3 0\nbogus\nr 2 3\nr 3 3\nr 3 2\n") == (
        "#1..\n#211\n####\n####\n"
        "#1..\n#211\n##F#\n####\n"
        "#1..\n1211\n.1F#\n.1##\n"
        "invalid command\n"
        "#1..\n1211\n.1F1\n.1##\n"
        "#1..\n1211\n.1F1\n.1#1\n"
        "#1..\n1211\n.1F1\n.111\n"
        "you win\n")

    # -- CLI: flag toggling then a mine ends the run with boom --
    assert run_cli(LAYOUT1, "r 0 0\nr 5 5\nf 1 1\nf 1 1\nr 2 2\n") == (
        "1#####\n######\n######\n######\n######\n######\n"
        "1#####\n######\n######\n######\n##1111\n##1...\n"
        "1#####\n#F####\n######\n######\n##1111\n##1...\n"
        "1#####\n######\n######\n######\n##1111\n##1...\n"
        "1#####\n#*####\n##*###\n####*#\n##1111\n#*1...\n"
        "boom\n")

    # -- CLI: junk, out-of-range and short commands are refused; q quits --
    assert run_cli(LAYOUT1, "x 1 1\nr 9 9\nr 1\n\nr 0 5\nq\nr 2 2\n") == (
        "invalid command\ninvalid command\ninvalid command\ninvalid command\n"
        "##1...\n##21..\n###211\n######\n######\n######\n")

    print("all minesweeper tests passed")


if __name__ == "__main__":
    main()

File: tasks/seeds/ts-minesweeper/files/test_minesweeper.ts
// Acceptance suite for the pure-function minesweeper engine.
// Run: node --test test_minesweeper.ts
//
// Boards load from fixture layout strings ('*' mine, '.' safe). Every
// mutator is pure: it returns a new Game and never modifies its argument;
// actions that change nothing return the exact same object.
import { test } from 'node:test';
import assert from 'node:assert/strict';
import {
  newGame,
  parseLayout,
  adjacency,
  reveal,
  mark,
  chord,
  flagsRemaining,
  render,
} from './minesweeper.ts';

// Two mines in opposite corners; every safe cell is one connected region.
const CORNERS = ['*....', '.....', '.....', '....*'].join('\n');
// Two separated mines for chord scenarios.
const CHORDY = ['.*...', '.....', '..*..'].join('\n');
// Tiny board where a win takes exactly two reveals.
const TINY = ['**', '..'].join('\n');

test('parseLayout reads a rectangular grid of . and *', () => {
  const mines = parseLayout('.*\n..\n');
  assert.deepEqual(mines, [[false, true], [false, false]]);
});

test('parseLayout rejects malformed layouts', () => {
  assert.throws(() => parseLayout(''), Error);
  assert.throws(() => parseLayout('\n\n'), Error);
  assert.throws(() => parseLayout('..\n.'), Error);
  assert.throws(() => parseLayout('.x\n..'), Error);
});

test('a fresh board renders fully hidden and is in play', () => {
  const g = newGame(CORNERS);
  assert.equal(g.status, 'playing');
  assert.equal(render(g), ['#####', '#####', '#####', '#####'].join('\n'));
  assert.equal(flagsRemaining(g), 2);
});

test('adjacency counts mines in the 8-neighborhood', () => {
  const g = newGame(CORNERS);
  assert.equal(adjacency(g, 0, 1), 1);
  assert.equal(adjacency(g, 1, 1), 1);
  assert.equal(adjacency(g, 3, 3), 1);
  assert.equal(adjacency(g, 1, 3), 0);
  const t = newGame(TINY);
  assert.equal(adjacency(t, 1, 0), 2);
  assert.equal(adjacency(t, 1, 1), 2);
});

test('coordinates outside the board throw RangeError', () => {
  const g = newGame(TINY);
  assert.throws(() => reveal(g, -1, 0), RangeError);
  assert.throws(() => reveal(g, 0, 2), RangeError);
  assert.throws(() => mark(g, 2, 0), RangeError);
  assert.throws(() => chord(g, 0, 5), RangeError);
  assert.throws(() => adjacency(g, 5, 0), RangeError);
});

test('revealing a numbered cell exposes only that cell', () => {
  const g = reveal(newGame(CORNERS), 0, 1);
  assert.equal(render(g), ['#1###', '#####', '#####', '#####'].join('\n'));
  assert.equal(g.status, 'playing');
});

test('revealing a zero cell flood-fills through the region', () => {
  const g = reveal(newGame(CORNERS), 1, 3);
  assert.equal(render(g), ['#1...', '11...', '...11', '...1#'].join('\n'));
});

test('reveal is pure: the original game is untouched', () => {
  const g0 = newGame(CORNERS);
  const before = render(g0);
  const g1 = reveal(g0, 1, 3);
  assert.notEqual(g1, g0);
  assert.equal(render(g0), before);
  assert.equal(g0.status, 'playing');
});

test('flood-fill never reveals a flagged cell', () => {
  let g = mark(newGame(CORNERS), 2, 2);
  g = reveal(g, 1, 3);
  assert.equal(render(g), ['#1...', '11...', '..F11', '...1#'].join('\n'));
  assert.equal(g.status, 'playing');
  assert.equal(flagsRemaining(g), 1);
});

test('flood-fill reveals question-marked cells and clears the mark', () => {
  let g = mark(mark(newGame(CORNERS), 2, 2), 2, 2);
  assert.equal(render(g).split('\n')[2], '##?##');
  g = reveal(g, 1, 3);
  assert.equal(render(g), ['#1...', '11...', '...11', '...1#'].join('\n'));
});

test('revealing a flagged cell is a no-op returning the same object', () => {
  const g = mark(newGame(CORNERS), 0, 0);
  assert.equal(reveal(g, 0, 0), g);
});

test('mark cycles none -> flag -> question -> none on hidden cells', () => {
  const g0 = newGame(CORNERS);
  const g1 = mark(g0, 1, 1);
  const g2 = mark(g1, 1, 1);
  const g3 = mark(g2, 1, 1);
  assert.equal(render(g1).split('\n')[1], '#F###');
  assert.equal(render(g2).split('\n')[1], '#?###');
  assert.equal(render(g3).split('\n')[1], '#####');
  assert.equal(flagsRemaining(g1), 1);
  assert.equal(flagsRemaining(g2), 2);
});

test('marking a revealed cell is a no-op returning the same object', () => {
  const g = reveal(newGame(CORNERS), 0, 1);
  assert.equal(mark(g, 0, 1), g);
});

test('a question-marked cell can still be revealed', () => {
  let g = mark(mark(newGame(CORNERS), 0, 1), 0, 1);
  g = reveal(g, 0, 1);
  assert.equal(render(g).split('\n')[0], '#1###');
});

test('flagsRemaining goes negative when overflagged', () => {
  let g = newGame(TINY);
  g = mark(g, 0, 0);
  g = mark(g, 0, 1);
  g = mark(g, 1, 0);
  assert.equal(flagsRemaining(g), -1);
});

test('revealing a mine loses; boom cell is !, other mines *', () => {
  const g = reveal(newGame(CORNERS), 3, 4);
  assert.equal(g.status, 'lost');
  assert.deepEqual(g.boom, [3, 4]);
  assert.equal(render(g), ['*####', '#####', '#####', '####!'].join('\n'));
});

test('lost render marks wrong flags X and keeps question marks hidden info', () => {
  let g = mark(newGame(CORNERS), 1, 1); // flag on a safe cell
  g = mark(mark(g, 0, 0), 0, 0); // question the (0,0) mine
  g = reveal(g, 3, 4);
  assert.equal(g.status, 'lost');
  assert.equal(render(g), ['*####', '#X###', '#####', '####!'].join('\n'));
});

test('after the game is over every action is a same-object no-op', () => {
  const lost = reveal(newGame(CORNERS), 3, 4);
  assert.equal(reveal(lost, 0, 1), lost);
  assert.equal(mark(lost, 1, 1), lost);
  assert.equal(chord(lost, 1, 1), lost);
});

test('revealing every safe cell wins regardless of flags', () => {
  let g = mark(newGame(TINY), 0, 0); // one correct flag, one mine unflagged
  g = reveal(g, 1, 0);
  assert.equal(g.status, 'playing');
  assert.equal(render(g), ['F#', '2#'].join('\n'));
  g = reveal(g, 1, 1);
  assert.equal(g.status, 'won');
  assert.equal(render(g), ['F#', '22'].join('\n'));
});

test('a single zero-region reveal can win outright', () => {
  const g = reveal(newGame(CORNERS), 1, 3);
  assert.equal(g.status, 'won');
});

test('chord on a satisfied number reveals the unflagged neighbors', () => {
  let g = reveal(newGame(CHORDY), 1, 0);
  assert.equal(render(g), ['#####', '1####', '#####'].join('\n'));
  g = mark(g, 0, 1);
  g = chord(g, 1, 0);
  assert.equal(g.status, 'playing');
  assert.equal(render(g), ['1F###', '12###', '.1###'].join('\n'));
});

test('chord with the wrong number of flags is a same-object no-op', () => {
  const g = reveal(newGame(CHORDY), 1, 0);
  assert.equal(chord(g, 1, 0), g); // zero flags placed
  const two = mark(mark(g, 0, 0), 0, 1); // two flags around a 1
  assert.equal(chord(two, 1, 0), two);
});

test('question marks do not count toward chord satisfaction', () => {
  let g = reveal(newGame(CHORDY), 1, 0);
  g = mark(mark(g, 0, 1), 0, 1); // question on the actual mine
  assert.equal(chord(g, 1, 0), g);
});

test('chord on a hidden cell or a zero cell is a same-object no-op', () => {
  const g0 = newGame(CHORDY);
  assert.equal(chord(g0, 1, 0), g0); // hidden
  const g1 = reveal(g0, 0, 3); // opens the zero region on the right
  assert.equal(g1.status, 'playing');
  assert.equal(chord(g1, 0, 4), g1); // revealed but adjacency 0
});

test('chording over a wrong flag hits the mine in reading order', () => {
  let g = reveal(newGame(CHORDY), 1, 0);
  g = mark(g, 0, 0); // flag the safe corner instead of the mine
  g = chord(g, 1, 0);
  assert.equal(g.status, 'lost');
  assert.deepEqual(g.boom, [0, 1]);
  assert.equal(render(g), ['X!###', '1####', '##*##'].join('\n'));
});

test('chord is pure: source position survives a losing chord', () => {
  let g = reveal(newGame(CHORDY), 1, 0);
  g = mark(g, 0, 0);
  const boom = chord(g, 1, 0);
  assert.equal(boom.status, 'lost');
  assert.equal(g.status, 'playing');
  assert.equal(render(g), ['F####', '1####', '#####'].join('\n'));
});

test('a chord can finish the board and win', () => {
  // Flag both mines on CHORDY, open the left block with a chord, then
  // finish the right side; the final board keeps its flags.
  let g = newGame(CHORDY);
  g = mark(g, 0, 1);
  g = mark(g, 2, 2);
  g = reveal(g, 1, 0);
  g = chord(g, 1, 0); // opens (0,0), (1,1) and floods (2,0)-(2,1)
  g = reveal(g, 0, 2);
  g = reveal(g, 1, 2);
  g = reveal(g, 0, 3); // zero: floods the whole right side
  assert.equal(g.status, 'won');
  assert.equal(render(g), ['1F1..', '1221.', '.1F1.'].join('\n'));
});
